src/NHibernate/Linq/DefaultQueryProvider.cs

```diff
@@ -299,6 +315,17 @@ public int ExecuteDml<T>(QueryMode queryMode, Expression expression)
 			return query.ExecuteUpdate();
 		}
 
+		public IAsyncEnumerable<T> GetAsyncEnumerable<T>(Expression expression)
+		{
+			var nhLinqExpression = PrepareQuery(expression, out var query);
+			if (nhLinqExpression.ExpressionToHqlTranslationResults?.PostExecuteTransformer != null)
+			{
+				throw new NotSupportedException("AsyncEnumerable is not supported when PostExecuteTransformer is set.");
```

Comment:
> Is this transformation for whole list?

Yes, the `PostExecuteTransformer` takes the whole list as an argument and it can be used for aggregate methods like `Enumerable.Count` and also for `Enumerable.All` to support polymorphic queries:
https://github.com/nhibernate/nhibernate-core/blob/ac39173567b31bcfad475ca32687c9faf0d37f87/src/NHibernate/Linq/Visitors/ResultOperatorProcessors/ProcessAll.cs#L25-L27

As `AsAsyncEnumerable` does not have an overload like `ToFutureValue` does, where you can do something like `q.ToFutureValue(o => o.Count())`, this code will never be reached when using the current `AsAsyncEnumerable` extension method. We could add an overload for it and cast the result to `IEnumerable<T>` as `AsyncEnumerableImpl` now implements it, but it kills the purpose of `IAsyncEnumerable` as the iteration will be done synchronously using `IEnumerable` instead.